### PR TITLE
Fix constant buffer being stomped on by shadows during async tasks

### PR DESCRIPTION
**Pull request by FrancescoC-unity**
This is a tad complex in that the global constant buffer is used by some tasks that run in the async compute pipe, but the shadow task that is running on the graphics pipe is updating the content of said global constant buffer. 

The CBuffer share the same memory so stomping happens; this PR fixes the issue by allowing a new instance of a constant buffer to be used by shadows. 

This required changing CBuffer API, also I changed how shadow atlas are init as it started to be very hard to read/follow. 

Fix for https://fogbugz.unity3d.com/f/cases/1324551/

NOTE: This _needs_ backport, but the backport is likely to be complex. I'll do a manual backport. I still targeted hd/bugfix, but let me know if you want me to target master instead. 

I'll wait for an answer to ^^ before doing the backport.

### CODE REVIEW

**Review by JulienIgnace-Unity — APPROVED** [review]

**Review by sebastienlagarde — APPROVED** [review]

**Comment by sebastienlagarde** [comment]
No QA requried, the Yamato test are enough

File: com.unity.render-pipelines.core/Runtime/Common/ConstantBuffer.cs [hunk 1]
```diff
@@ -122,86 +122,134 @@ public static void ReleaseAll()
...
+
+        ComputeBuffer m_GPUConstantBuffer = null;
+
+        /// <summary>
+        /// Constant Buffer constructor.
+        /// </summary>
+        public ConstantBuffer()
+        {
+            m_GPUConstantBuffer = new ComputeBuffer(1, UnsafeUtility.SizeOf<CBType>(), ComputeBufferType.Constant);
+            ConstantBuffer.Register(this);
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.core/Runtime/Common/ConstantBuffer.cs:164]: Register() should only be called for Singleton constant buffers (as it's necessary for ReleaseAll to work).
Explicitly allocated CBs must be released explicitly as well.

File: com.unity.render-pipelines.core/Runtime/Common/ConstantBuffer.cs [hunk 2]
```diff
@@ -122,86 +122,134 @@ public static void ReleaseAll()
...
-
-                CoreUtils.SafeRelease(m_GPUConstantBuffer);
-                s_Instance = null;
+                s_Instance = value;
             }
         }
+
+        public override void Release()
+        {
+            s_Instance = null;
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.core/Runtime/Common/ConstantBuffer.cs:252]: Need to call base.Release() here

File: com.unity.render-pipelines.core/Runtime/Common/ConstantBuffer.cs [hunk 3]
```diff
@@ -105,14 +105,14 @@ public class ConstantBuffer
...
-            var cb = TypedConstantBuffer<CBType>.instance;
+            var cb = ConstantBufferSingleton<CBType>.instance;
 
             cb.Set(mat, shaderId);
         }
 
         /// <summary>
         /// Release all currently allocated constant buffers.
-        /// This needs to be called before shutting down the application.
+        /// This needs to be called before shutting down the application.
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.core/Runtime/Common/ConstantBuffer.cs:115]: Maybe Update the doc to explain it's only for CBs used through the static API? (not sure of the right wording here)

File: com.unity.render-pipelines.core/Runtime/Common/ConstantBuffer.cs [hunk 4]
```diff
@@ -19,7 +19,7 @@ public class ConstantBuffer
         /// <param name="shaderId">Shader porperty id to bind the constant buffer to.</param>
         public static void PushGlobal<CBType>(CommandBuffer cmd, in CBType data, int shaderId) where CBType : struct
```
Comment (JulienIgnace-Unity) [inline com.unity.render-pipelines.core/Runtime/Common/ConstantBuffer.cs:20]: Might be worth adding the PushGlobal version to regular CBs as well for consistency.

**Pull request merged**